php-checks/src/main/java/org/sonar/php/checks/NullDereferenceInConditionalCheck.java

```diff
@@ -0,0 +1,121 @@
+import org.sonar.plugins.php.api.visitors.CheckContext;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+@Rule(
+  key = "S1697",
+  name = "Short-circuit logic should be used to prevent null pointer dereferences in conditionals",
+  priority = Priority.BLOCKER,
+  tags = {Tags.BUG})
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("2min")
+public class NullDereferenceInConditionalCheck extends PHPVisitorCheck {
+
+  private static final String MESSAGE_FORMAT =
+    "Either reverse the equality operator in the \"%s\" null test, or reverse the logical operator that follows it.";
+
+  @Override
+  public void visitBinaryExpression(BinaryExpressionTree tree) {
+    if (isAndWithEqualToNull(tree) || isOrWithNonEqualToNull(tree)) {
+      BinaryExpressionTree leftOperand = (BinaryExpressionTree) removeParenthesis(tree.leftOperand());
+      ExpressionTree expression = removeParenthesis(getNonNullOperand(leftOperand));
+      tree.rightOperand().accept(new NullExpressionUsageVisitor(expression, context(), this));
+    }
+
+    super.visitBinaryExpression(tree);
+  }
+
+  private static ExpressionTree getNonNullOperand(BinaryExpressionTree binaryExpressionTree) {
+    if (isNull(binaryExpressionTree.leftOperand())) {
+      return binaryExpressionTree.rightOperand();
+    }
+    return binaryExpressionTree.leftOperand();
+  }
+
+  private static boolean isAndWithEqualToNull(BinaryExpressionTree tree) {
+    return tree.is(Tree.Kind.CONDITIONAL_AND, Kind.ALTERNATIVE_CONDITIONAL_AND)
+      && isNullComparison(tree.leftOperand(), Tree.Kind.EQUAL_TO, Tree.Kind.STRICT_EQUAL_TO);
+  }
+
+  private static boolean isOrWithNonEqualToNull(BinaryExpressionTree tree) {
+    return tree.is(Tree.Kind.CONDITIONAL_OR, Kind.ALTERNATIVE_CONDITIONAL_OR)
+      && isNullComparison(tree.leftOperand(), Tree.Kind.NOT_EQUAL_TO, Tree.Kind.STRICT_NOT_EQUAL_TO);
+  }
+
+  private static boolean isNullComparison(ExpressionTree expression, Tree.Kind kind1, Tree.Kind kind2) {
+    ExpressionTree tree = removeParenthesis(expression);
+    if (tree.is(kind1, kind2)) {
+      BinaryExpressionTree binaryExp = (BinaryExpressionTree) tree;
+      return isNull(binaryExp.leftOperand()) || isNull(binaryExp.rightOperand());
+    }
+    return false;
+  }
+
+  private static ExpressionTree removeParenthesis(ExpressionTree expressionTree) {
+    if (expressionTree.is(Kind.PARENTHESISED_EXPRESSION)) {
+      return removeParenthesis(((ParenthesisedExpressionTree) expressionTree).expression());
+    }
+    return expressionTree;
+  }
+
+  private static boolean isNull(Tree tree) {
+    return tree.is(Tree.Kind.NULL_LITERAL);
+  }
+
+  private class NullExpressionUsageVisitor extends PHPVisitorCheck {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2694)